### PR TITLE
Therese view establishment

### PR DESCRIPTION
**Pull request by TheresePetersson**
Added interface function of Predicates and Action of the View Establishment module. Check ENUM for "No action", "Reset" and "No return value".

### CODE REVIEW

File: modules/view_establishment/predicates.py [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+        )
+        return("Reset")
+
+    def view_change(self):
+        """A view change is required."""
+        self.vChange = True
+
+    def get_view(self, node_j):
+        """Returns the most recent reported view of node_j."""
+        if (node_j == self.id or
```
Comment (axelniklasson) [inline modules/view_establishment/predicates.py]: Should be `if (node_jd == self.id) and ...`

File: modules/view_establishment/predicates.py [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+        if (phase == 0):
+            return self.automation_phase_0(type, case)
+
+        # Phase 1, in a view change
+        elif(phase == 1):
+            return self.automation_phase_1(type, case)
+
+        # Not a valid phase
+        else:
+            raise Exception('Not a valid phase: {}'.format(phase))
```
Comment (axelniklasson) [inline modules/view_establishment/predicates.py]: Would be nice to raise [ValueError](https://docs.python.org/2/library/exceptions.html#exceptions.ValueError) instead, will make it easier later to have as specific exceptions as possible

File: modules/view_establishment/predicates.py [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+            # Adopt the new view
+            if(case == 0):
+                if(self.view_pair_to_adopt):
+                    self.adopt(self.view_pair_to_adopt)
+                    self.view_pair_to_adopt = None
+                    self.view_module.next_phs()
+                    self.reset_v_change()
+                    return "Adopted new view"
+                else:
+                    raise Exception("Not a valid view pair to adopt")
```
Comment (axelniklasson) [inline modules/view_establishment/predicates.py]: Same here, ValueError

File: modules/view_establishment/predicates.py [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+                # TODO What do return?
+
+            # No action and reset the v_change-variable
+            elif(case == 2):
+                self.reset_v_change()
+                return "No action"
+
+            # Full reset
+            elif(case == 3):
+                return self.reset_all()
```
Comment (axelniklasson) [inline modules/view_establishment/predicates.py:248]: Should call `reset_v_change` as well, maybe add a TODO here to remember to add call to `reset_v_change` when it is implemented

File: modules/view_establishment/predicates.py [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+
+            # Adopt the new transit view
+            if(case == 0):
+                if(self.view_pair_to_adopt):
+                    self.adopt(self.view_pair_to_adopt)
+                    self.view_pair_to_adopt = None
+                    self.reset_v_change()
+                    return "Adopted new view"
+                else:
+                    raise Exception("Not a valid view pair to adopt")
```
Comment (axelniklasson) [inline modules/view_establishment/predicates.py]: ValueError

File: modules/view_establishment/predicates.py [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+                return self.transit_adopble(self.id, 1, "Remain")
+
+            # True if there is no adoptable view (predicates for other cases
+            # are false)
+            elif(case == 3):
+                return True
+
+            # Not a valid case
+            else:
+                raise Exception('Not a valid case: {}'.format(case))
```
Comment (axelniklasson) [inline modules/view_establishment/predicates.py]: ValueError

File: modules/view_establishment/predicates.py [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+                self.reset_v_change()
+                return "No action"
+
+            # Full reset
+            elif(case == 3):
+                return self.reset_all()
+
+            # Not a valid case
+            else:
+                raise Exception('Not a valid case: {}'.format(case))
```
Comment (axelniklasson) [inline modules/view_establishment/predicates.py]: ValueError

File: modules/view_establishment/predicates.py [hunk 8]
```diff
@@ -0,0 +1,344 @@
...
+
+        # Not a valid type (act or pred)
+        else:
+            raise Exception('Not a valid type: {}'.format(type))
```
Comment (axelniklasson) [inline modules/view_establishment/predicates.py]: ValueError

File: modules/view_establishment/predicates.py [hunk 9]
```diff
@@ -0,0 +1,344 @@
...
+                self.reset_v_change()
+                return "No action"
+
+            # Reset all
+            elif(case == 3):
+                return self.reset_all()
+
+            # Not a valid case
+            else:
+                raise Exception('Not a valid case: {}'.format(case))
```
Comment (axelniklasson) [inline modules/view_establishment/predicates.py]: ValueError

File: modules/view_establishment/predicates.py [hunk 10]
```diff
@@ -0,0 +1,344 @@
...
+
+        # Not a valid type (act or pred)
+        else:
```
Comment (axelniklasson) [inline modules/view_establishment/predicates.py:332]: ValueError

File: modules/view_establishment/predicates.py [hunk 11]
```diff
@@ -0,0 +1,344 @@
...
+            raise Exception('Not a valid type: {}'.format(type))
+
+    def auto_max_case(self, phase):
+        """Returns the max case for the phase."""
+        return 3 if phase == 0 else 3
```
Comment (axelniklasson) [inline modules/view_establishment/predicates.py:337]: ¯\\_(ツ)_/¯

**Review by axelniklasson — COMMENTED** [review]
Awesome work! Minor changes that needs to be addressed and then it's ready to merge!

File: modules/view_establishment/predicates.py (hunk 1 above)
Comment (TheresePetersson) [reply modules/view_establishment/predicates.py]: Done, updated test accordingly

File: modules/view_establishment/predicates.py (hunk 2 above)
Comment (TheresePetersson) [reply modules/view_establishment/predicates.py]: Done

File: modules/view_establishment/predicates.py (hunk 3 above)
Comment (TheresePetersson) [reply modules/view_establishment/predicates.py]: Done

File: modules/view_establishment/predicates.py (hunk 4 above)
Comment (TheresePetersson) [reply modules/view_establishment/predicates.py:248]: Added the call and updated corresponding test

File: modules/view_establishment/predicates.py (hunk 5 above)
Comment (TheresePetersson) [reply modules/view_establishment/predicates.py]: Done

File: modules/view_establishment/predicates.py (hunk 7 above)
Comment (TheresePetersson) [reply modules/view_establishment/predicates.py]: Done

File: modules/view_establishment/predicates.py (hunk 8 above)
Comment (TheresePetersson) [reply modules/view_establishment/predicates.py]: Done

File: modules/view_establishment/predicates.py (hunk 10 above)
Comment (TheresePetersson) [reply modules/view_establishment/predicates.py:332]: Done

File: modules/view_establishment/predicates.py (hunk 11 above)
Comment (TheresePetersson) [reply modules/view_establishment/predicates.py:337]: Insane logic.

File: modules/view_establishment/predicates.py (hunk 9 above)
Comment (TheresePetersson) [reply modules/view_establishment/predicates.py]: Done

File: modules/view_establishment/predicates.py (hunk 6 above)
Comment (TheresePetersson) [reply modules/view_establishment/predicates.py]: Done

**Review by axelniklasson — APPROVED** [review]

**Pull request merged**